packages/analytics/index.test.ts

```diff
@@ -34,51 +34,144 @@ import { GtagCommand, EventName } from './src/constants';
+        AnalyticsError.NO_API_KEY
+      );
+    });
+    it('Warns if config has no apiKey but does have a measurementId', () => {
+      stubFetch(200, { measurementId: fakeMeasurementId });
+      const warnStub = stub(console, 'warn');
+      const app = getFakeApp({
+        appId: fakeAppParams.appId,
+        measurementId: fakeMeasurementId
+      });
+      const installations = getFakeInstallations();
+      analyticsFactory(app, installations);
+      expect(warnStub.args[0][1]).to.include(fakeMeasurementId);
+      expect(warnStub.args[0][1]).to.include('Falling back to');
+      warnStub.restore();
+      fetchStub.restore();
+    });
+    it('Throws if cookies are not enabled', () => {
+      const cookieStub = stub(navigator, 'cookieEnabled').value(false);
+      const app = getFakeApp({
+        appId: fakeAppParams.appId,
+        apiKey: fakeAppParams.apiKey
+      });
+      const installations = getFakeInstallations();
+      expect(() => analyticsFactory(app, installations)).to.throw(
+        AnalyticsError.COOKIES_NOT_ENABLED
+      );
+      cookieStub.restore();
+    });
+    it('Throws if browser extension environment', () => {
+      window.chrome = { runtime: { id: 'blah' } };
+      const app = getFakeApp({
+        appId: fakeAppParams.appId,
+        apiKey: fakeAppParams.apiKey
+      });
+      const installations = getFakeInstallations();
+      expect(() => analyticsFactory(app, installations)).to.throw(
+        AnalyticsError.INVALID_ANALYTICS_CONTEXT
+      );
+      window.chrome = undefined;
+    });
+    it('Throws if indexedDB does not exist', () => {
+      const idbStub = stub(window, 'indexedDB').value(undefined);
+      const app = getFakeApp({
+        appId: fakeAppParams.appId,
+        apiKey: fakeAppParams.apiKey
+      });
+      const installations = getFakeInstallations();
+      expect(() => analyticsFactory(app, installations)).to.throw(
+        AnalyticsError.INDEXED_DB_UNSUPPORTED
+      );
+      idbStub.restore();
+    });
+    it('Warns eventually if indexedDB.open() does not work', async () => {
+      stubFetch(200, { measurementId: fakeMeasurementId });
+      const clock = useFakeTimers();
+      const warnStub = stub(console, 'warn');
+      const idbOpenStub = stub(indexedDB, 'open').throws(
+        'idb open throw message'
+      );
+      const app = getFakeApp({
+        appId: fakeAppParams.appId,
+        apiKey: fakeAppParams.apiKey
+      });
+      const installations = getFakeInstallations();
+      analyticsFactory(app, installations);
+      await clock.runAllAsync();
+      expect(warnStub.args[0][1]).to.include(
+        AnalyticsError.INVALID_INDEXED_DB_CONTEXT
+      );
+      expect(warnStub.args[0][1]).to.include('idb open throw message');
+      warnStub.restore();
+      idbOpenStub.restore();
+      fetchStub.restore();
+    });
+    it('Throws if creating an instance with already-used analytics ID', () => {
```

Comment:
Is it more accurate to say `already-used appId`, since we are checking if the appId is already initialized?

Reply:
Done.